subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/FixedSharedModeCrossProcessCacheAccess.java

```diff
@@ -64,27 +64,37 @@ public void open() {
         try {
             boolean rebuild = initializationAction.requiresInitialization(fileLock);
             if (rebuild) {
+                Exception latestException = null;
                 for (int tries = 0; rebuild && tries < 3; tries++) {
                     fileLock.close();
                     fileLock = null;
-                    final FileLock exclusiveLock = lockManager.lock(lockTarget, lockOptions.withMode(Exclusive), cacheDisplayName);
+                    FileLock exclusiveLock = null;
+                    try {
+                        exclusiveLock = lockManager.lock(lockTarget, lockOptions.withMode(Exclusive), cacheDisplayName);
+                    } catch (Exception e) {
+                        // acquiring the exclusive lock can fail in the rare case where another process is just doing or has just done the cache initialization
+                        latestException = e;
+                    }
                     try {
                         rebuild = initializationAction.requiresInitialization(exclusiveLock);
```

Comment:
:o: I'm wondering if this could cause an NPE for "some initialization action" not aware tha `exclusiveLock` can now be null?

Reply:
Ahh. Good catch. I totally missed that.

No we can not call it without lock. That's why further down a new shared lock is acquired before calling `requiresInitialization()` again. And that's in fact enough, because the loop will terminate if that call succeeds. We do not need this addition `rebuild = initializationAction.requiresInitialization(exclusiveLock)` here. I removed it.